File: src/agt_bt_executor/test/test_execute_waypoint_action.py
from pathlib import Path

ROOT = Path(__file__).parents[1]
SOURCE = (ROOT / "src/execute_waypoint_task_action.cpp").read_text()
HEADER = (ROOT / "include/agt_bt_executor/execute_waypoint_task_action.hpp").read_text()
ACTION = (ROOT.parent / "agt_interfaces/action/ExecuteWaypointTask.action").read_text()


def test_uses_formal_project_waypoint_action_and_typed_ports():
    assert "agt_interfaces/action/execute_waypoint_task.hpp" in HEADER
    assert '"/agt/navigation/execute_waypoint_task"' in SOURCE
    for port in ("map_id", "map_version_id", "task_group_id", "task_revision",
                 "expected_content_sha256", "loop_count", "client_request_id"):
        assert f'"{port}"' in SOURCE


def test_deprecated_goal_inputs_are_not_populated():
    assert "task_file" in ACTION and "poses" in ACTION and "bool loop" in ACTION
    goal_builder = SOURCE.split("bool ExecuteWaypointTask::makeGoal", 1)[1]
    assert "task_file" not in goal_builder
    assert "poses" not in goal_builder
    assert "goal.loop =" not in goal_builder

File: src/agt_bt_executor/test/test_execute_waypoint_action.cpp
#include <gtest/gtest.h>
#include <agt_interfaces/action/execute_waypoint_task.hpp>
#include <behaviortree_cpp/blackboard.h>
#include <rclcpp/rclcpp.hpp>
#include <rclcpp_action/rclcpp_action.hpp>
#include <thread>
#include "agt_bt_executor/execute_waypoint_task_action.hpp"

TEST(ExecuteWaypointTaskBtAction, FakeServerReceivesTypedGoal) {
  using Action = agt_interfaces::action::ExecuteWaypointTask;
  rclcpp::init(0, nullptr); auto node = std::make_shared<rclcpp::Node>("waypoint_bt_test"); Action::Goal received;
  auto server = rclcpp_action::create_server<Action>(node, "/agt/navigation/execute_waypoint_task",
    [&received](const rclcpp_action::GoalUUID &, std::shared_ptr<const Action::Goal> goal) { received = *goal; return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE; },
    [](std::shared_ptr<rclcpp_action::ServerGoalHandle<Action>>) { return rclcpp_action::CancelResponse::ACCEPT; },
    [](std::shared_ptr<rclcpp_action::ServerGoalHandle<Action>> handle) { std::thread([handle]() { std::this_thread::sleep_for(std::chrono::milliseconds(500)); auto result = std::make_shared<Action::Result>(); result->success = true; handle->succeed(result); }).detach(); });
  rclcpp::executors::SingleThreadedExecutor executor; executor.add_node(node);
  BT::NodeConfiguration config; config.input_ports["map_id"] = "map"; config.input_ports["map_version_id"] = "v1"; config.input_ports["task_group_id"] = "rows"; config.input_ports["task_revision"] = "7"; config.input_ports["expected_content_sha256"] = "sha256:abc"; config.input_ports["loop_count"] = "2"; config.input_ports["client_request_id"] = "req-1"; config.input_ports["timeout_s"] = "2.0";
  agt_bt_executor::ExecuteWaypointTask action("waypoint", config, node); ASSERT_EQ(action.executeTick(), BT::NodeStatus::RUNNING);
  BT::NodeStatus status = BT::NodeStatus::RUNNING;
  for (int i = 0; i < 300; ++i) { executor.spin_some(); status = action.executeTick(); if (status != BT::NodeStatus::RUNNING) break; std::this_thread::sleep_for(std::chrono::milliseconds(10)); }
  EXPECT_EQ(status, BT::NodeStatus::SUCCESS); EXPECT_EQ(received.map_id, "map"); EXPECT_EQ(received.task_revision, 7U); EXPECT_EQ(received.loop_count, 2U); EXPECT_TRUE(received.task_file.empty()); EXPECT_TRUE(received.poses.empty());
  rclcpp::shutdown();
}
